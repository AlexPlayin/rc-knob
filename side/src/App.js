import React from 'react'

import { H1, H2, Grid, Description } from './styled'
import PointerExample from './examples/Pointer'
import ScaleExample from './examples/Scale'
import ArcExample from './examples/Arc'
import ArcPointerExample from './examples/Arc_Pointer'
import ScalePointerExample from './examples/Scale_Pointer'
import ValueExample from './examples/Value'
import ArcPointerValueExample from './examples/Arc_Pointer_Value'
import ScaleScale from './examples/Scale_Scale'
import PointerCustom from './examples/Pointer_custom'
import ScaleCustom from './examples/Scale_custom'
import ScaleStyled from './examples/Scale_styled'
import ScaleStyled2 from './examples/Scale_styled_2'
import ArcPointerStyled from './examples/Arc_Pointer_styled'
import EventMouse from "./examples/Event_mouse";
import EventReadOnly from "./examples/Event_readOnly";
import EventNoWheel from "./examples/Event_nowheel";


import Filter from './filter'

export default () => {
    return (
        <div>
            <Filter />
            <H1>React Knob</H1>
            <H2>
                <span>Base Components</span>
            </H2>
            <Description>
                The base of all knobs is the <pre>Knob</pre> component which handles
                all of the user interaction and calculation of the knob value. The
                user can interact via drag'n drop, mousewheel and keyboard. The
                knobs are accessible by the <pre>tab</pre> key.
                <br /> The display part is splitted into the <pre>
                    Pointer
                </pre>, <pre>Scale</pre>, <pre>Arc</pre> and <pre>Value</pre>. One
                or more of them can be <pre>children</pre> of a <pre>Knob</pre>.
            </Description>
            <Grid>
                <PointerExample />
                <ArcExample />
                <ScaleExample />
                <ValueExample />
            </Grid>
            <H2>
                <span>Combinations</span>
            </H2>
            <Description>
                You can combine all of the base components in a knob.
            </Description>
            <Grid>
                <ArcPointerExample />
                <ScalePointerExample />
                <ArcPointerValueExample />
                <ScaleScale />
            </Grid>
            <H2>
                <span>Custom Elements</span>
            </H2>
            <Description>
                The <pre>Pointer</pre> and <pre>Scale</pre> components can also
                render custom components.
            </Description>
            <Grid>
                <PointerCustom />
                <ScaleCustom />
            </Grid>
            <H2>
                <span>Events</span>
            </H2>
            <Grid>
                <EventMouse />
                <EventReadOnly />
                <EventNoWheel />
            </Grid>
            <H2>
                <span>Styles</span>
            </H2>
            <Description>
                The components and the knob itself can be styled by addingthe{' '}
                <pre>className</pre> prop. The <pre>Scale</pre> can have an
                additional <pre>activeClassname</pre> prop to style the active tick
                differently. As all components are SVG you can also use filters in
                your styles.
            </Description>
            <Grid>
                <ArcPointerStyled />
                <ScaleStyled />
                <ScaleStyled2 />
            </Grid>
        </div>
    )
}
