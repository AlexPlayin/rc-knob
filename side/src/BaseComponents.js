import React from 'react'

import { H1, H2, Grid, Description } from './styled'
import PointerExample from './examples/Pointer'
import ScaleExample from './examples/Scale'
import ArcExample from './examples/Arc'
import ArcPointerExample from './examples/Arc_Pointer'
import ScalePointerExample from './examples/Scale_Pointer'
import ValueExample from './examples/Value'
import ArcPointerValueExample from './examples/Arc_Pointer_Value'
import ScaleScale from './examples/Scale_Scale'
import PointerCustom from './examples/Pointer_custom'
import ScaleCustom from './examples/Scale_custom'

export default () => (
    <div>
        <H1>React Knob</H1>
        <H2>
            <span>Base Components</span>
        </H2>
        <Description>
            The base of all knobs is the <pre>Knob</pre> component which handles
            all of the user interaction and calculation of the knob value.{' '}
            <br /> The display part is splitted into the <pre>Pointer</pre>,{' '}
            <pre>Scale</pre>, <pre>Arc</pre> and <pre>Value</pre>. One or more{' '}
            of them can be <pre>children</pre> of a <pre>Knob</pre>.
        </Description>
        <Grid>
            <PointerExample />
            <ArcExample />
            <ScaleExample />
            <ValueExample />
        </Grid>
        <H2>
            <span>Combinations</span>
        </H2>
        <Description>
            You can combine all of the base components in a knob.
        </Description>
        <Grid>
            <ArcPointerExample />
            <ScalePointerExample />
            <ArcPointerValueExample />
            <ScaleScale />
        </Grid>
        <H2>
            <span>Custom Elements</span>
        </H2>
        <Description>
            The <pre>Knob</pre> component and the <pre>Pointer</pre> and{' '}
            <pre>Scale</pre> components can also render custom components. All
            direct children of the <pre>Knob</pre> get the following props
            passed: <pre>percentage</pre> <pre>size</pre>, <pre>value</pre>,
            <pre>angleOffset</pre>,<pre>angleRange</pre>,{' '}
            <pre>radius: size / 2</pre>,<pre>center: size / 2</pre>,{' '}
            <pre>steps</pre>. The children of <pre>Pointer</pre> get{' '}
            <pre>width</pre>, <pre>height</pre>, <pre>percentage</pre> passed.
        </Description>
        <Grid>
            <PointerCustom />
            <ScaleCustom />
        </Grid>
    </div>
)
