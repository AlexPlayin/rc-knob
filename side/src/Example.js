import React from 'react'
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter'
import { H3, Cell, CodeBlock } from './styled'
import { Knob } from 'rc-knob'
import theme from './prismTheme'

const renderKnob = (child, options = {}) => (
    <Knob
        onChange={value => console.log(value)}
        size={100}
        value={0}
        angleOffset={220}
        angleRange={280}
        min={0}
        max={100}
        {...options}
        snap
    >
        {child}
    </Knob>
)

export default ({ example, knob, title, options }) => (
    <Cell>
        <H3>{title}</H3>
        {renderKnob(knob, options)}
        <CodeBlock>
            <SyntaxHighlighter language="jsx" style={theme}>
                {example}
            </SyntaxHighlighter>
        </CodeBlock>
    </Cell>
)
