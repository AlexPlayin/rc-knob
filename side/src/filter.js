import React from 'react'

export default () => (
    <svg>
        <filter id="inner-shadow">
            <feOffset dx="0" dy="5" />
            <feGaussianBlur stdDeviation="5" result="offset-blur" />
            <feComposite
                operator="out"
                in="SourceGraphic"
                in2="offset-blur"
                result="inverse"
            />
        </filter>
    </svg>
)
