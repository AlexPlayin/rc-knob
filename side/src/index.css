@import url('https://fonts.googleapis.com/css?family=Work+Sans:100,300');
@import url('https://fonts.googleapis.com/css?family=Eczar:800');

html {
    height: 100%;
}

body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
        'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
        'Helvetica Neue', sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    height: 100%;
    padding-top: 254px;
}

code {
    font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
        monospace;
}
.value {
    font-family: 'Eczar', serif;
    fill: #fc5a96;
    font-size: 50px;
}
.value2 {
    font-family: 'Work Sans', sans-serif;
    fill: #fc5a96;
    font-size: 30px;
}
