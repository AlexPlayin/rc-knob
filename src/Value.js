import React from 'react'

export const Value = ({
    value,
    size,
    decimalPlace = 0,
    className,
    marginBottom = 0,
}) => (
    <text
        style={{ userSelect: 'none' }}
        x="50%"
        textAnchor="middle"
        className={className}
        y={size - marginBottom}
    >
        {value.toFixed(decimalPlace)}
    </text>
)
