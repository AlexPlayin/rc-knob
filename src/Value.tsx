import React from 'react'
import type { PropsWithKnobState } from 'types';

interface Props {
    decimalPlace?: number;
    className: string;
    marginBottom?: number;
}

export const Value = ({
    value,
    size,
    decimalPlace = 0,
    className,
    marginBottom = 0,
}: PropsWithKnobState<Props>) => {
    if (value === null || value === undefined) {
        return <></>
    }
    let label = value.toFixed(decimalPlace)
    // make sure no negative zero is displayed
    if (label.startsWith("-") && Number.parseFloat(label) === 0) {
        label = label.slice(1)
    }
    return (
        <text
            style={{ userSelect: 'none' }}
            x="50%"
            textAnchor="middle"
            className={className}
            y={(size ?? 0) - marginBottom}
        >
            {label}
        </text>
    )
}
