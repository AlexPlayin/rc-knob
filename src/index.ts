export * from './Knob'
export * from './Arc'
export * from './Scale'
export * from './Pointer'
export * from './Value'
export * from './Range'
export * from './Spiral'
export * from './Label'
export * from './types'
