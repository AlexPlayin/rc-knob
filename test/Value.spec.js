import React from 'react'
import { shallow } from 'enzyme'
import { Value } from '../src/Value'

describe('Value', () => {
    it('renders correct with a given value', () => {
        const component = shallow(
            <Value size={50} value={10.1212} className="someClassName" />
        )
        expect(component).toMatchSnapshot()
    })
    it('renders nothing correct without a given value', () => {
        const component = shallow(<Value size={50} className="someClassName" />)
        expect(component.html()).toBe(null)
    })
})
