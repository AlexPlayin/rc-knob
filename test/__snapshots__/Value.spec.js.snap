// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value renders correct with a given value 1`] = `
<text
  className="someClassName"
  style={
    Object {
      "userSelect": "none",
    }
  }
  textAnchor="middle"
  x="50%"
  y={50}
>
  10
</text>
`;
